<a href="https://colab.research.google.com/github/fowardelcac/Titanic_2.0/blob/main/Modelado_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
! git clone https://github.com/fowardelcac/Titanic_2.0.git

fatal: destination path 'Titanic_2.0' already exists and is not an empty directory.


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import random
import warnings

from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

warnings.filterwarnings('ignore')

# Funciones

In [3]:
def split(df):
  n = len(df.columns) - 1
  X = df.drop('Survived', axis=1).values.reshape((-1, n))
  y = df.Survived
  X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.3, random_state = 42)
  return X_train, X_test, y_train, y_test

In [4]:
def modelar(modelo, X_train, y_train):
   return modelo.fit(X_train, y_train) 

In [5]:
def predecir(modelo, X_test, y_test, accuracy_score):
  y_pred = modelo.predict(X_test)
  score = round(accuracy_score(y_test, y_pred), 3)
  return score

def predecir_rdo(modelo, X_test):
  return modelo.predict(X_test)

In [6]:
def edit_age(df):
  serie = df['Age'].dropna()
  # Obtener la cantidad de valores nulos en la columna 'Age'
  size = df['Age'].isna().sum()
  # Generar una nueva Serie con valores aleatorios de la Serie 'serie'
  random_values = np.random.choice(serie, size=size)
  # Reemplazar los valores NaN en la columna 'Age' con los valores aleatorios
  df.loc[df['Age'].isna(), 'Age'] = random_values
  return df

In [7]:
def onehot(df):
  encoder = OneHotEncoder()
  data = pd.DataFrame(encoder.fit_transform(df['Pclass'].values.reshape(-1, 1)).toarray(), columns=['1', '2', '3'])
  return data

In [8]:
def edit(df):
  df = edit_age(df)

  mapeo = {'male': 0, 'female': 1}
  df.Sex = df.Sex.map(mapeo)
  
  df['familia'] = df.SibSp + df.Parch

  data = onehot(df)
  df.reset_index('PassengerId', inplace=True)
  df = pd.concat([df, data], axis=1).set_index('PassengerId')

  scaler = MinMaxScaler().fit_transform(df['Fare'].values.reshape(-1, 1))
  scaled_series = pd.Series(scaler.flatten(), index=df.index)
  df.Fare = scaled_series

  df.drop(['SibSp', 'Parch', 'Cabin', 'Ticket', 'Name', 'Pclass', 'Embarked'], axis = 1, inplace = True)
  return df   

In [9]:
def modelar_():
  modelos_lista = [RandomForestClassifier(), DecisionTreeClassifier(), KNeighborsClassifier(), LogisticRegression()]
  data = create_df_azar(df)
  cols_d = data.columns[:-1]
  
  X_train, X_test, y_train, y_test = split(data)
  dic_m = {}
  for i in modelos_lista:  
    modelo = modelar(i, X_train, y_train)
    acc = predecir(modelo, X_test, y_test, accuracy_score)
    dic_m[i] = [cols_d, acc]
  return dic_m

In [10]:
def random_samples():
  columnas = ['Sex', 'Age', 'Fare', 'familia', '1', '2', '3']
  n_col =  random.randint(1, len(columnas))
  random_c = random.sample(columnas, n_col)
  return random_c

def create_df_azar(df):
  columnas = random_samples()
  data = df.filter(columnas, axis = 1)
  data['Survived'] = df.Survived
  return data

In [11]:
def grid(modelo, params, X_train, y_train):
  grid_rf = GridSearchCV(modelo, params, cv = 5).fit(X_train, y_train)
  mejores_parametros = grid_rf.best_params_
  mejor_score = grid_rf.best_score_

  print("Mejores parámetros:", mejores_parametros)
  print("Mejor score:", mejor_score)

In [12]:
def rdo(test, y_pred):
  return pd.DataFrame(
    {
    'Survived': y_pred
    }, index= test.index )
   

In [13]:
df = pd.read_csv('/content/Titanic_2.0/Datasets_originales/train.csv').set_index('PassengerId')
df = edit(df)
df.head()

Survived  Sex   Age      Fare  familia    1    2    3
PassengerId                                                       
1                   0    0  22.0  0.014151        1  0.0  0.0  1.0
2                   1    1  38.0  0.139136        1  1.0  0.0  0.0
3                   1    1  26.0  0.015469        0  0.0  0.0  1.0
4                   1    1  35.0  0.103644        1  1.0  0.0  0.0
5                   0    0  35.0  0.015713        0  0.0  0.0  1.0

In [ ]:
scaler = MinMaxScaler().fit_transform(df['Age'].values.reshape(-1, 1))
scaled_series = pd.Series(scaler.flatten(), index=df.index)
df.Age = scaled_series


In [ ]:
X_train, X_test, y_train, y_test = split(df)

In [ ]:
modelos_lista = [RandomForestClassifier(), DecisionTreeClassifier(), KNeighborsClassifier(), LogisticRegression()]

Sex	Age	Fare	familia	1	2	3 age y fare escaladas



In [ ]:
for i in modelos_lista:
  modelo = modelar(i, X_train, y_train)
  acc = predecir(modelo, X_test, y_test, accuracy_score)
  print(f"Modelo: {i}, score:", acc)

Modelo: RandomForestClassifier(), score: 0.792
Modelo: DecisionTreeClassifier(), score: 0.74
Modelo: KNeighborsClassifier(), score: 0.808
Modelo: LogisticRegression(), score: 0.8


In [ ]:
lista_dic =[]
for i in range(15):
  lista_dic.append(modelar_())

In [ ]:
lista_dic[0].keys()

dict_keys([RandomForestClassifier(), DecisionTreeClassifier(), KNeighborsClassifier(), LogisticRegression()])

In [ ]:
scores = []
for diccionario in lista_dic:
    for item in diccionario.values():
        scores.append(item[1])

for index, diccionario in enumerate(lista_dic):
    for clave in diccionario.keys():
        print(index+1, clave, ':', diccionario[clave][1])
        print('*' * 50)

1 RandomForestClassifier() : 0.646
1 DecisionTreeClassifier() : 0.646
1 KNeighborsClassifier() : 0.657
1 LogisticRegression() : 0.715
2 RandomForestClassifier() : 0.788
2 DecisionTreeClassifier() : 0.734
2 KNeighborsClassifier() : 0.801
2 LogisticRegression() : 0.798
3 RandomForestClassifier() : 0.771
3 DecisionTreeClassifier() : 0.769
3 KNeighborsClassifier() : 0.766
3 LogisticRegression() : 0.79
4 RandomForestClassifier() : 0.792
4 DecisionTreeClassifier() : 0.737
4 KNeighborsClassifier() : 0.785
4 LogisticRegression() : 0.792
5 RandomForestClassifier() : 0.652
5 DecisionTreeClassifier() : 0.614
5 KNeighborsClassifier() : 0.684
5 LogisticRegression() : 0.707
6 RandomForestClassifier() : 0.609
6 DecisionTreeClassifier() : 0.609
6 KNeighborsClassifier() : 0.609
6 LogisticRegression() : 0.609
7 RandomForestClassifier() : 0.644
7 DecisionTreeClassifier() : 0.633
7 KNeighborsClassifier() : 0.662
7 LogisticRegression() : 0.668
8 RandomForestClassifier() : 0.758
8 DecisionTreeClassifier() :

In [ ]:
lista_dic[10]

{RandomForestClassifier(): [Index(['familia', 'Sex', '3', 'Fare'], dtype='object'),
  0.79],
 DecisionTreeClassifier(): [Index(['familia', 'Sex', '3', 'Fare'], dtype='object'),
  0.768],
 KNeighborsClassifier(): [Index(['familia', 'Sex', '3', 'Fare'], dtype='object'),
  0.792],
 LogisticRegression(): [Index(['familia', 'Sex', '3', 'Fare'], dtype='object'),
  0.801]}


1. RF:  '3', 'Fare', '2', 'Sex', 'Age', 'familia', '1', -> 0.798 / 'familia', 'Sex', '3', 'Fare' -> 0.79
2. Arbol:  '3', 'Fare', '2', 'Sex', 'Age', 'familia', '1', -> 0.756 / 'Sex', '3', -> 0.787
3. KNN:   '3', 'Fare', '2', 'Sex', 'Age', 'familia', '1', -> 0.808 / 'Fare', 'familia', 'Sex', '2', 'Age', '3' -> 0.801
4. LR:   'familia', 'Sex', '3', 'Fare', -> 0.801


# Random forest

In [39]:
data_rf = df.filter([ 'familia', 'Sex', '3', 'Fare', 'Survived'], axis = 1)
X_train, X_test, y_train, y_test = split(data_rf)
modelo_rf = modelar(RandomForestClassifier(), X_train, y_train)
acc = predecir(modelo_rf, X_test, y_test, accuracy_score)
print(acc)

0.784


In [40]:
parametros = {
    'n_estimators': [100, 200, 300],  # número de árboles en el bosque
    'max_depth': [None, 5, 10],  # profundidad máxima de cada árbol
    'min_samples_split': [2, 5, 10],  # número mínimo de muestras requeridas para dividir un nodo
    'min_samples_leaf': [1, 2, 4]  # número mínimo de muestras requeridas en cada hoja
}

grid(RandomForestClassifier(), parametros, X_train, y_train)

Mejores parámetros: {'max_depth': 5, 'min_samples_leaf': 2, 'min_samples_split': 10, 'n_estimators': 300}
Mejor score: 0.8167714884696018


In [46]:
modelo_rf = modelar(RandomForestClassifier(max_depth =  5, min_samples_leaf = 2, min_samples_split = 10, n_estimators = 300), X_train, y_train)
acc = predecir(modelo_rf, X_test, y_test, accuracy_score)
print(acc)

0.803


# Arbol de desicion

In [15]:
data_a = df.filter([ '3', 'Fare', '2', 'Sex', 'Age', 'familia', '1', 'Survived'], axis = 1)
X_train, X_test, y_train, y_test = split(data_a)
modelo_tree = modelar(DecisionTreeClassifier(), X_train, y_train)
acc = predecir(modelo_tree, X_test, y_test, accuracy_score)
print(acc)

0.728


In [29]:
parametros = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10, 15],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}
grid(DecisionTreeClassifier(), parametros, X_train, y_train)

Mejores parámetros: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 4, 'min_samples_split': 2}
Mejor score: 0.8352201257861636


In [32]:
modelo_tree = modelar(DecisionTreeClassifier(criterion = 'gini', max_depth = None, min_samples_leaf = 4, min_samples_split= 2), X_train, y_train)
acc = predecir(modelo_tree, X_test, y_test, accuracy_score)
print(acc)

0.79


# KNN

In [16]:
data_knn = df.filter(['Fare', 'familia', 'Sex', '2', 'Age', '3', 'Survived'], axis = 1)
X_train, X_test, y_train, y_test = split(data_knn)
modelo_knn = modelar(KNeighborsClassifier(), X_train, y_train)
acc = predecir(modelo_knn, X_test, y_test, accuracy_score)
print(acc)

0.692


# Regresion logistica

In [17]:
data_rl = df.filter(['familia', 'Sex', '3', 'Fare', 'Survived'], axis = 1)
X_train, X_test, y_train, y_test = split(data_rl)
modelo_rl = modelar(LogisticRegression(), X_train, y_train)
acc = predecir(modelo_rl, X_test, y_test, accuracy_score)
print(acc)

0.801


In [22]:
parametros = {
    'C': [0.1, 1.0, 10.0],
     'penalty': ['l1', 'l2'],
    'solver': ['liblinear', 'saga'],
    'fit_intercept': [True, False],
     'class_weight': [None, 'balanced'],
     'max_iter': [100, 500, 1000]
}
grid(LogisticRegression(), parametros, X_train, y_train)

Mejores parámetros: {'C': 1.0, 'class_weight': None, 'fit_intercept': False, 'max_iter': 100, 'penalty': 'l1', 'solver': 'liblinear'}
Mejor score: 0.7979734451432565


In [24]:
modelo_rl = modelar(LogisticRegression(C = 1.0, class_weight =  None, fit_intercept = False, max_iter = 100, penalty = 'l1', solver= 'liblinear'), X_train, y_train)
acc = predecir(modelo_rl, X_test, y_test, accuracy_score)
print(acc)

0.809


1. Arbol de desicion: 0.812
2. Regresion: 0.809
3. Random fores: 0.801

# Test

In [17]:
test = pd.read_csv('/content/Titanic_2.0/Datasets_originales/test.csv').set_index('PassengerId')
test = edit(test)
test.Fare.fillna(test.Fare.mean(), inplace=True)

In [72]:
test.isnull().sum()

Sex        0
Age        0
Fare       0
familia    0
1          0
2          0
3          0
dtype: int64

### Ajusto los tres modelos

1. RL: 0.809
2. RF: 0.803 
3. Tree: 0.8

In [53]:
modelo_rl = modelar(LogisticRegression(C = 1.0, class_weight =  None, fit_intercept = False, max_iter = 100, penalty = 'l1', solver= 'liblinear'), X_train, y_train)
acc = predecir(modelo_rl, X_test, y_test, accuracy_score)
print(acc)

0.809


In [76]:
X_test = test.filter(['familia', 'Sex', '3', 'Fare'], axis = 1)

In [78]:
test_rdo_rl = rdo(test, predecir_rdo(modelo_rl, X_test))
test_rdo_rl.to_csv('Reg_log2.csv')

In [49]:
modelo_rf = modelar(RandomForestClassifier(max_depth =  5, min_samples_leaf = 2, min_samples_split = 10, n_estimators = 300), X_train, y_train)
acc = predecir(modelo_rf, X_test, y_test, accuracy_score)
print(acc)

0.803


In [79]:
X_test = test.filter(['familia', 'Sex', '3', 'Fare'], axis=1)
test_rdo_rf = rdo(test, predecir_rdo(modelo_rf, X_test))
test_rdo_rf.to_csv('Rf2.csv')

In [16]:
modelo_tree = modelar(DecisionTreeClassifier(criterion = 'gini', max_depth = None, min_samples_leaf = 4, min_samples_split= 2), X_train, y_train)
acc = predecir(modelo_tree, X_test, y_test, accuracy_score)
print(acc)

0.771


In [83]:
X_test = test.filter(['3', 'Fare', '2', 'Sex', 'Age', 'familia', '1'], axis=1)
X_test

3      Fare    2  Sex   Age  familia    1
PassengerId                                             
892          1.0  0.015282  0.0    0  34.5        0  0.0
893          1.0  0.013663  0.0    1  47.0        1  0.0
894          0.0  0.018909  1.0    0  62.0        0  0.0
895          1.0  0.016908  0.0    0  27.0        0  0.0
896          1.0  0.023984  0.0    1  22.0        2  0.0
...          ...       ...  ...  ...   ...      ...  ...
1305         1.0  0.015713  0.0    0  48.0        0  0.0
1306         0.0  0.212559  0.0    1  39.0        0  1.0
1307         1.0  0.014151  0.0    0  38.5        0  0.0
1308         1.0  0.015713  0.0    0  18.0        0  0.0
1309         1.0  0.043640  0.0    0  26.0        2  0.0

[418 rows x 7 columns]

In [19]:
X_test = test.filter(['3', 'Fare', '2', 'Sex', 'Age', 'familia', '1'], axis=1)
test_rdo_tree = rdo(test, predecir_rdo(modelo_tree, X_test))
test_rdo_tree.to_csv('arbol2.csv')